Both points are right and I have taken both. Thank you for actually running the test against a base checkout rather than reading it, because that is what caught the real problem.

**The `#[should_panic]` test proved nothing.** On the vulnerable base the `debug_assert!` supplies the panic, so it passed there for a completely different reason than it passed on the patch. A green test carrying no information is worse than no test, and you demonstrated that rather than asserting it.

**Malformed bytes now cross the `Result` boundary.** `parse_length`, `unzip` and `new` are fallible. A truncated prefix returns a contextual `Error::corrupt_file_named` naming the width and what remained:

```
truncated length prefix: 4 byte(s) remain in the page buffer but a 64-bit
length prefix requires 8
```

`create_decoder` already returned `Result`, so this propagates with a single `?` at the one production call site. No panic path is introduced.

## On the regression test discriminating

I owe you a precise answer rather than a claim.

The replacement asserts the error variant and the message text, across both supported prefix widths (a 3 byte buffer for a 32 bit prefix, a 4 byte buffer for a 64 bit prefix):

```rust
let err = decode_variable_full_zip(vec![0xAA; buf_len], bits)
    .expect_err("a truncated length prefix must not decode");
assert!(matches!(err, Error::CorruptFile { .. }), ...);
assert!(err.to_string().contains("truncated length prefix"), ...);
```

On the vulnerable base this test **cannot compile**, which is a stronger form of discrimination than failing at runtime. I checked it the way you did, applying only the test hunk to a detached base checkout:

```
error[E0308]: mismatched types
    --> rust/lance-encoding/src/encodings/logical/primitive.rs:9882:9
9878 |  ) -> lance_core::Result<super::VariableFullZipDecoder> {
     |       ------------------------------------------------- expected `Result<...>` because of return type
9882 |  super::VariableFullZipDecoder::new(
```

`new` returns `Self` on the base, so a test that asserts an error cannot be written against it at all. The test exercises a contract the vulnerable code does not have, and there is therefore no way for it to pass on unpatched code.

If you would prefer a test that also demonstrates the underlying read on the base, that has to be a sanitizer run rather than a unit test, since on the base the failure is undefined behaviour rather than a value. For reference, that is how the original report was produced: a release build with `-Zsanitizer=address` gives `heap-buffer-overflow, READ of size 8, 3 bytes after a 4 byte region` on the base and no sanitizer error after this change, with the well-formed control unaffected in both.

## Verification

- both new tests pass on the patch
- the base checkout with only the test hunk fails to compile, as above
- the crate's existing suite is unchanged

Happy to adjust the error message wording or the naming if you would like something that matches your conventions more closely.

--- a/rust/lance-encoding/src/encodings/logical/primitive.rs
+++ b/rust/lance-encoding/src/encodings/logical/primitive.rs
@@ -9866,9 +9866,10 @@ mod tests {
         use crate::repdef::{ControlWordParser, DefinitionInterpretation};
         use std::sync::Arc;
         Arc::new(super::FullZipDecodeDetails {
-            value_decompressor: super::PerValueDecompressor::Variable(
-                Arc::new(VariableDecoder::default()) as Arc<dyn VariablePerValueDecompressor>,
-            ),
+            value_decompressor: super::PerValueDecompressor::Variable(Arc::new(
+                VariableDecoder::default(),
+            )
+                as Arc<dyn VariablePerValueDecompressor>),
             def_meaning: vec![DefinitionInterpretation::NullableItem].into(),
             ctrl_word_parser: ControlWordParser::new(0, 0),
             max_rep: 0,

